### ISSUE TITLE
Determine the safest way to use the testing sendcloud service

### ISSUE THREAD

**Issue by Alienpruts** [haringsrob/Sendcloud-Drupal-Commerce#5]
This still bothers me :

https://github.com/haringsrob/Sendcloud-Drupal-Commerce/commit/85671741b2e2d821e9c266fd3fb23fff7b58b798

especially these lines :

https://github.com/haringsrob/Sendcloud-Drupal-Commerce/blob/85671741b2e2d821e9c266fd3fb23fff7b58b798/commerce_sendcloud.module#L377

(off topic, how do i get code snippets in these issues?)

I really haven't had the time to further look into this, but do you know of a way to safely determine the correct id? Hardcoding is pretty dangerous : suppose that id suddenly becomes a paying Sendcloud method and people who have checked the test checkbox are suddenly finding their Sendcloud accounts being billed. I'd be really mad :)

**Comment by haringsrob** [comment]
Hi @Alienpruts,

I asked @SendCloud for a test environment or a mock but they cannot provide, so what I want to suggest is the following, that we send a shipping request for each of the shipping methods, and save the requests in our own mock.

Then we could setup simpletests or behat to use that mock to test sendcloud. Or make a submodule that contains the mock.

**Comment by Alienpruts** [comment]
Sounds like a plan, @haringsrob .

However, time constraints and my lack of operative knowledge of testing frameworks (I really should get into this) mean I cannot help you as much as I'd like on this?

However, objectively speaking : you'd do a request for all shipping methods, save them (mock them) and decide which one is safe to use for testing (ie. doesn't have any costs assigned to them)?  Something like that?

**Comment by haringsrob** [comment]
No, The idea is to do the actual request to Sendcloud, with a specific set of data. 

ex: Create parcel on my address using method X

Then the response I receive (json or xml) I save to a custom file. And I cancel the order on the back office of Sendcloud.

Then I put all of these json/xml documents in a folder, and instead of sending my request to Sendcloud, I will use a custom menu hook, that will translate my test request and return the contents of one of the files. 

This way we can work offline, and test faster without stuffing the Sendcloud back office.

I will create the first examples tonight, and show how they work :), but the idea is to be able to work on this offline.

**Comment by Alienpruts** [comment]
Ah, I see now.

Nice plan, but one caveat maybe : wouldn't you be restricted to performing the exact same request over and over? Since you downloaded the response bodies for those specific requests, that means any 'generic' way of testing is out of the question.

Don't misunderstand me, it is a great idea for getting started on testing, but I believe we will restrict ourselves in the long run to those specific requests (and responses), which is not always ideal. Suppose that SendCloud introduces new shipping methods without us knowing about it, we would have to resend the request(s) in order to obtain the new responses, do you see? :)

If it would only be used for us devs for testing, then indeed we wouldn't have to worry to much about it. But if you would use this idea for expanding the Testing part of the service, I believe we will have to think some more about it :)

Anywhere I can help, shout. I'll try to fit it into my schedule.

**Comment by haringsrob** [comment]
That's true, but it's the best I can come up with as they don't provide any testing services. I'll try to contact one of their developers for more information.

**Comment by Alienpruts** [comment]
:+1:
